File: backend/jventure/regen.jl
require("exp.jl")

function regenAndAttach(trace::Trace,border::Array{Node},scaffold::Scaffold,shouldRestore::Bool,db::DB)
  weight = 0.0
  for node = border
    if isAbsorbing(scaffold,node)
      weight += attach(trace,node,scaffold,shouldRestore,db)
    else
      weight += regen(trace,node,scaffold,shouldRestore,db)
      if isObservation(trace,node)
        weight += constrain(trace,node,getObservedValue(trace,node))
      end
    end
  end
  return weight
end

constrain(trace::Trace,node::LookupNode,value::Any) = constrain(trace,node.sourceNode,value)
function constrain(trace::Trace,node::Node,value::Any)
  if isa(getPSP(trace,node),ESRRefOutputPSP)
    return constrain(trace,getESRSourceNode(trace,node),value)
  else
    (psp,args) = (getPSP(trace,node),getArgs(trace,node))
    oldValue = getGroundValue(trace,node)
    unincorporate!(psp,oldValue,args)
    weight = logDensity(psp,value,args)
    setValue!(trace,node,value)
    incorporate!(psp,value,args)
    setConstrained!(trace,node)
    unregisterRandomChoice!(trace,node)
    return weight
  end
end

function attach(trace::Trace,node::ApplicationNode,scaffold::Scaffold,shouldRestore::Bool,db::DB)
  weight = regenParents(trace,node,scaffold,shouldRestore,db)
  (psp::PSP,args::Args,value) = (getPSP(trace,node),getArgs(trace,node),getGroundValue(trace,node))
  weight += logDensity(psp,value,args)
  incorporate!(psp,value,args)
  return weight
end

function regenParents(trace::Trace,node::Node,scaffold::Scaffold,shouldRestore::Bool,db::DB)
  weight = 0.0
  for parent = getParents(trace,node)
    weight += regen(trace,parent,scaffold,shouldRestore,db)
  end
  return weight
end

function regen(trace::Trace,node::Node,scaffold::Scaffold,shouldRestore::Bool,db::DB)
  weight = 0.0
  if isResampling(scaffold,node)
    if getRegenCount(scaffold,node) == 0
      weight += regenParents(trace,node,scaffold,shouldRestore,db)
      if isa(node,LookupNode)
        setValue!(trace,node,getValue(trace,node.sourceNode))
      else
        weight += applyPSP(trace,node,scaffold,shouldRestore,db)
        if isa(node,RequestNode)
          weight += evalRequests(trace,node,scaffold,shouldRestore,db)
        end
      end
    end
    incrementRegenCount!(scaffold,node)
  end

  value = getValue(trace,node)
  if isa(value,SPRef) && value.makerNode != node && isAAA(scaffold,value.makerNode)
    weight += regen(trace,value.makerNode,scaffold,shouldRestore,db)
  end

  return weight
end


function evalFamily(trace::Trace,sym::Symbol,env::ExtendedEnvironment,scaffold::Scaffold,db::DB)
  sourceNode = findSymbol(env,sym)
  regen(trace,sourceNode,scaffold,false,db)
  return (0,createLookupNode(trace,sourceNode))
end

function evalFamily(trace::Trace,value::VentureValue,env::ExtendedEnvironment,scaffold::Scaffold,db::DB)
  return (0,createConstantNode(trace,value))
end

function evalFamily(trace::Trace,exp::Array{VentureValue},env::ExtendedEnvironment,scaffold::Scaffold,db::DB)
  @assert !isempty(exp)
  if isQuotation(exp)
    return (0,createConstantNode(trace,textOfQuotation(exp)))
  else
    (weight::Float64,operatorNode::Node) = evalFamily(trace,getOperator(exp),env,scaffold,db)
    operands = getOperands(exp)
    operandNodes = Array(Node,length(operands))
    for i in 1:length(operands)
      (w::Float64,operandNodes[i]) = evalFamily(trace,operands[i],env,scaffold,db)
      weight += w
    end

    (requestNode,outputNode) = createApplicationNodes(trace,operatorNode,operandNodes,env)
    weight += apply(trace,requestNode,outputNode,scaffold,false,db)
    return weight,outputNode
  end
end

function apply(trace::Trace,requestNode::RequestNode,outputNode::OutputNode,scaffold::Scaffold,shouldRestore::Bool,db::DB)
  weight = applyPSP(trace,requestNode,scaffold,shouldRestore,db)
  weight += evalRequests(trace,requestNode,scaffold,shouldRestore,db)
  weight += applyPSP(trace,outputNode,scaffold,shouldRestore,db)
  return weight
end

function processMadeSP(trace::Trace,node::Union(ConstantNode,OutputNode),isAAA::Bool)
  sp = getValue(trace,node)
  @assert isa(sp,SP)
  if !isAAA
    createMadeSPRecord!(trace,node,sp,constructSPAux(sp.outputPSP))
    if hasAEKernel(sp.outputPSP) registerAEKernel!(trace,node) end
  else
    setMadeSP!(trace,node,sp)
  end
  setValue!(trace,node,SPRef(node))
end

function applyPSP(trace::Trace,node::ApplicationNode,scaffold::Scaffold,shouldRestore::Bool,db::DB)
  (psp,args) = (getPSP(trace,node),getArgs(trace,node))
  weight = 0.0

  if hasValue(db,node)
    oldValue = getValue(db,node)
  else
    oldValue = nothing
  end

  if shouldRestore
    @assert hasValue(db,node)
    newValue = oldValue
  elseif hasLKernel(scaffold,node)
    k = getLKernel(scaffold,node)
    newValue = ksimulate(k,trace,oldValue,args)
    weight += kweight(k,trace,newValue,oldValue,args)
  else
    newValue = simulate(psp,args)
  end

  @assert newValue != nothing
  setValue!(trace,node,newValue)
  incorporate!(psp,newValue,args)

  @assert !node.valid
  node.valid = true

  if isa(newValue,SP) processMadeSP(trace,node,isAAA(scaffold,node)) end
  if isRandom(psp) registerRandomChoice!(trace,node) end
  if isa(psp,ApplyInScopeOutputPSP)
    operandNodes = getOperandNodes(trace,node)
    (scope,block) = [getValue(trace,n) for n in operandNodes[1:2]]
    blockNode = operandNodes[3]
    @assert isRandom(getPSP(trace,blockNode))
    registerRandomChoiceInScope!(trace,scope,block,blockNode)
  end
  return weight
end

function evalRequests(trace::Trace,node::RequestNode,scaffold::Scaffold,shouldRestore::Bool,db::DB)
  weight = 0.0
  request = getValue(trace,node)

  spRecord = getSPRecord(trace,node)
  (sp,aux,families) = (spRecord.sp,spRecord.aux,spRecord.families)

  # first evaluate exposed simulation requests (ESRs)
  for esr = request.esrs
    if !haskey(families,esr.id)
      if shouldRestore
        esrParent = getSPFamilyRoot(db,sp,esr.id)
        weight += restore(trace,esrParent,scaffold,db)
      else
        (w,esrParent) = evalFamily(trace,esr.exp,esr.env,scaffold,db)
        weight += w
      end
      families[esr.id] = esrParent
    else
      esrParent = families[esr.id]
      weight += regen(trace,esrParent,scaffold,shouldRestore,db)
    end
    esrParent = families[esr.id]
    registerESREdge!(trace,esrParent,getOutputNode(trace,node))
  end
  return weight
end


#TODO

function restore(trace::Trace,node::Node,scaffold::Scaffold,db::DB)
  if isa(node,ConstantNode)
    return 0
  elseif isa(node,LookupNode)
    weight = regen(trace,node.sourceNode,scaffold,true,db)
    setValue!(trace,node,getValue(trace,node.sourceNode))
    reconnectLookup!(trace,node)
    return weight
  else # node is output node
    weight = restore(trace,getOperatorNode(trace,node),scaffold,db)
    for operandNode = getOperandNodes(trace,node)
      weight += restore(trace,operandNode,scaffold,db)
    end
    weight += apply(trace,node.requestNode,node,scaffold,true,db)
    return weight
  end
end

File: backend/new_cxx/src/regen.cxx
#include "regen.h"

#include "scaffold.h"
#include "trace.h"
#include "node.h"
#include "expressions.h"
#include "env.h"
#include "sp.h"
#include "db.h"
#include "psp.h"
#include "lkernel.h"
#include "srs.h"
#include "sps/scope.h"
#include <iostream>
#include <boost/foreach.hpp>


using std::cout;
using std::endl;

double regenAndAttach(Trace * trace,
                      const vector<Node*> & border,
                      shared_ptr<Scaffold> scaffold,
                      bool shouldRestore,
                      shared_ptr<DB> db,
                      shared_ptr<map<Node*,Gradient> > gradients)
{
  double weight = 0;
  map<Node*,VentureValuePtr> constraintsToPropagate;
  for (size_t i = 0; i < border.size(); ++i)
  {
    Node * node = border[i];
    if (scaffold->isAbsorbing(node))
    {
      ApplicationNode * appNode = dynamic_cast<ApplicationNode*>(node);
      assert(appNode);
      weight += attach(trace,appNode,scaffold,shouldRestore,db,gradients);
    }
    else
    {
      weight += regen(trace,node,scaffold,shouldRestore,db,gradients);
      if (trace->isObservation(node))
      {
        OutputNode * outputNode = trace->getConstrainableNode(node);
        weight += constrain(trace,outputNode,trace->getObservedValue(node));
        constraintsToPropagate[outputNode] = trace->getObservedValue(node);
      }
    }
  }
  // Propagate constraints
  for (map<Node*,VentureValuePtr>::iterator iter1 = constraintsToPropagate.begin();
       iter1 != constraintsToPropagate.end();
       ++iter1)
  {
    Node * node = iter1->first;
    set<Node*> children = trace->getChildren(node);
    for (set<Node*>::iterator iter2 = children.begin();
         iter2 != children.end();
         ++iter2)
    {
      propagateConstraint(trace,*iter2,iter1->second);
    }
  }
  return weight;
}


double constrain(Trace * trace,
                 OutputNode * node,
                 VentureValuePtr value)
{
  shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(node))->getPSP(node);
  shared_ptr<Args> args = trace->getArgs(node);

  psp->unincorporate(trace->getValue(node),args);
  double weight = psp->logDensity(value,args);
  trace->setValue(node,value);
  psp->incorporate(value,args);
  trace->registerConstrainedChoice(node);
  return weight;
}


void propagateConstraint(Trace * trace,
                         Node * node,
                         VentureValuePtr value)
{
  LookupNode * lookupNode = dynamic_cast<LookupNode*>(node);
  RequestNode * requestNode = dynamic_cast<RequestNode*>(node);
  OutputNode * outputNode = dynamic_cast<OutputNode*>(node);
  if (lookupNode) { trace->setValue(lookupNode,value); }
  else if (requestNode)
  {
    shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(requestNode))->getPSP(requestNode);
    if (!dynamic_pointer_cast<NullRequestPSP>(psp)) { throw "Cannot make requests downstream of a node that gets constrained during regen"; }
  }
  else
  {
    assert(outputNode);
    shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(outputNode))->getPSP(outputNode);
    if (psp->isRandom()) { throw "Cannot make random choices downstream of a node that gets constrained during regen"; }
    trace->setValue(node,psp->simulate(trace->getArgs(outputNode),trace->getRNG()));
  }
  set<Node*> children = trace->getChildren(node);
  for (set<Node*>::iterator iter = children.begin();
       iter != children.end();
       ++iter)
  {
    propagateConstraint(trace,*iter,value);
  }
}


double attach(Trace * trace,
              ApplicationNode * node,
              shared_ptr<Scaffold> scaffold,
              bool shouldRestore,
              shared_ptr<DB> db,
              shared_ptr<map<Node*,Gradient> > gradients)
{
  //cout << "attach(" << node << ")" << endl;

  double weight = regenParents(trace,node,scaffold,shouldRestore,db,gradients);
  shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(node))->getPSP(node);
  shared_ptr<Args> args = trace->getArgs(node);
  VentureValuePtr groundValue = trace->getGroundValue(node);
  weight += psp->logDensity(groundValue,args);
  psp->incorporate(groundValue,args);

  if (dynamic_pointer_cast<ScopeIncludeOutputPSP>(psp))
  {
    ScopeID scope = trace->getValue(node->operandNodes[0]);
    BlockID block = trace->getValue(node->operandNodes[1]);
    Node * blockNode = node->operandNodes[2];
    trace->registerUnconstrainedChoiceInScope(scope,block,blockNode);
  }    

  return weight;
}


double regen(Trace * trace,
              Node * node,
              shared_ptr<Scaffold> scaffold,
              bool shouldRestore,
              shared_ptr<DB> db,
              shared_ptr<map<Node*,Gradient> > gradients)
{ 
  //cout << "regenOuter(" << node << ")" << endl;
  double weight = 0;
  if (scaffold->isResampling(node))
  {
    //cout << "regen(" << node << ") = " << trace->getRegenCount(scaffold,node) << endl;
    if (trace->getRegenCount(scaffold,node) == 0)
    {
      weight += regenParents(trace,node,scaffold,shouldRestore,db,gradients);
      LookupNode * lookupNode = dynamic_cast<LookupNode*>(node);
      RequestNode * requestNode = dynamic_cast<RequestNode*>(node);
      OutputNode * outputNode = dynamic_cast<OutputNode*>(node);
      if (lookupNode) { trace->setValue(node, trace->getValue(lookupNode->sourceNode)); }
      else if (requestNode)
      {
        weight += applyPSP(trace,requestNode,scaffold,shouldRestore,db,gradients);
        weight += evalRequests(trace,requestNode,scaffold,shouldRestore,db,gradients);
      }
      else
      {
        assert(outputNode);
        weight += applyPSP(trace,outputNode,scaffold,shouldRestore,db,gradients);
      }
    }
    trace->incRegenCount(scaffold,node);
  }
  VentureValuePtr value = trace->getValue(node);

  shared_ptr<VentureSPRef> spRef = dynamic_pointer_cast<VentureSPRef>(value);
  if (spRef && spRef->makerNode != node && scaffold->isAAA(spRef->makerNode))
  {
    weight += regen(trace,spRef->makerNode,scaffold,shouldRestore,db,gradients);
  }
  return weight;
}

double regenParents(Trace * trace,
              Node * node,
              shared_ptr<Scaffold> scaffold,
              bool shouldRestore,
              shared_ptr<DB> db,
              shared_ptr<map<Node*,Gradient> > gradients)
{
  double weight = 0;
  vector<Node*> definiteParents = node->getDefiniteParents();
  for (size_t i = 0; i < definiteParents.size(); ++i) { weight += regen(trace,definiteParents[i],scaffold,shouldRestore,db,gradients); }
  return weight + regenESRParents(trace,node,scaffold,shouldRestore,db,gradients);
}

double regenESRParents(Trace * trace,
              Node * node,
              shared_ptr<Scaffold> scaffold,
              bool shouldRestore,
              shared_ptr<DB> db,
              shared_ptr<map<Node*,Gradient> > gradients)
{
  double weight = 0;
  vector<RootOfFamily> esrRoots = trace->getESRParents(node);
  for (size_t i = 0; i < esrRoots.size(); ++i) { weight += regen(trace,esrRoots[i].get(),scaffold,shouldRestore,db,gradients); }
  return weight;
}

pair<double,Node*> evalFamily(Trace * trace,
                              VentureValuePtr exp,
                              shared_ptr<VentureEnvironment> env,
                              shared_ptr<Scaffold> scaffold,
                              bool shouldRestore,
                              shared_ptr<DB> db,
                              shared_ptr<map<Node*,Gradient> > gradients)
{
  if (isVariable(exp))
  {
    double weight = 0;
    shared_ptr<VentureSymbol> symbol = dynamic_pointer_cast<VentureSymbol>(exp);
    Node * sourceNode = env->lookupSymbol(symbol);
    weight = regen(trace,sourceNode,scaffold,shouldRestore,db,gradients);

    return make_pair(weight,trace->createLookupNode(sourceNode,exp));
  }
  else if (isSelfEvaluating(exp)) { return make_pair(0,trace->createConstantNode(exp)); }
  else if (isQuotation(exp)) { return make_pair(0,trace->createConstantNode(textOfQuotation(exp))); }
  else
  {
    assert(exp->hasArray());
    vector<VentureValuePtr> array = exp->getArray();
    pair<double,Node*> p = evalFamily(trace,array[0],env,scaffold,shouldRestore,db,gradients);
    double weight = p.first;
    Node * operatorNode = p.second;

    /* DEBUG */
    shared_ptr<VentureSPRef> spRef = dynamic_pointer_cast<VentureSPRef>(trace->getValue(operatorNode));
    assert(spRef);
    /* END DEBUG */

    vector<Node*> operandNodes;
    for (size_t i = 1; i < array.size(); ++i)
    {
      pair<double,Node*>p = evalFamily(trace,array[i],env,scaffold,shouldRestore,db,gradients);
      weight += p.first;
      operandNodes.push_back(p.second);
    }

    pair<RequestNode*,OutputNode*> appNodes = trace->createApplicationNodes(operatorNode,operandNodes,env,exp);
    weight += apply(trace,appNodes.first,appNodes.second,scaffold,shouldRestore,db,gradients);
    return make_pair(weight,appNodes.second);
  }
}

double apply(Trace * trace,
              RequestNode * requestNode,
             OutputNode * outputNode,
              shared_ptr<Scaffold> scaffold,
              bool shouldRestore,
              shared_ptr<DB> db,
              shared_ptr<map<Node*,Gradient> > gradients)
{
  double weight = applyPSP(trace,requestNode,scaffold,shouldRestore,db,gradients);
  weight += evalRequests(trace,requestNode,scaffold,shouldRestore,db,gradients);

  /* DEBUG */
  const vector<ESR>& esrs = trace->getValue(requestNode)->getESRs();
  assert(trace->getESRParents(outputNode).size() == esrs.size());  
  /* END DEBUG */

  weight += regenESRParents(trace,outputNode,scaffold,shouldRestore,db,gradients);
  weight += applyPSP(trace,outputNode,scaffold,shouldRestore,db,gradients);

  return weight;
}



void processMadeSP(Trace * trace, Node * makerNode, bool isAAA, bool shouldRestore, shared_ptr<DB> db)
{
  shared_ptr<VentureSPRecord> spRecord = dynamic_pointer_cast<VentureSPRecord>(trace->getValue(makerNode));
  assert(spRecord);
  shared_ptr<SP> sp = spRecord->sp;

  if (shouldRestore && db->hasMadeSPAux(makerNode)) { spRecord->spAux = db->getMadeSPAux(makerNode); }
  if (sp->hasAEKernel()) { trace->registerAEKernel(makerNode); }
  trace->setMadeSPRecord(makerNode,spRecord);
  if (isAAA) 
    { 
      OutputNode * outputNode = dynamic_cast<OutputNode*>(makerNode);
      assert(outputNode);
      trace->discardAAAMadeSPAux(outputNode); 
    }
  trace->setValue(makerNode,shared_ptr<VentureValue>(new VentureSPRef(makerNode)));
}

double applyPSP(Trace * trace,
              ApplicationNode * node,
              shared_ptr<Scaffold> scaffold,
              bool shouldRestore,
              shared_ptr<DB> db,
              shared_ptr<map<Node*,Gradient> > gradients)
{
  //cout << "applyPSP(" << node << ")" << endl;
  double weight = 0;
  shared_ptr<PSP> psp = trace->getMadeSP(trace->getOperatorSPMakerNode(node))->getPSP(node);
  shared_ptr<Args> args = trace->getArgs(node);

  VentureValuePtr oldValue;
  VentureValuePtr newValue;

  if (db->hasValue(node)) { oldValue = db->getValue(node); }

  if (trace->hasLKernel(scaffold,node))
  {
    shared_ptr<LKernel> k = trace->getLKernel(scaffold,node);
    if (shouldRestore) { newValue = oldValue; }
    else { newValue = k->simulate(trace,oldValue,args,trace->getRNG()); }

    weight += k->weight(trace,newValue,oldValue,args);
    /*
      These lines were causing problems, due to a mismatch between vector<double>
      and double in the gradientOfLogDensity. Puma doesn't actually have variational
      though, so I'm commenting these out for now.
      
    shared_ptr<VariationalLKernel> vk = dynamic_pointer_cast<VariationalLKernel>(k);
    if (vk) { 
      assert(gradients);
      gradients->insert(make_pair(node,vk->gradientOfLogDensity(newValue,args))); 
    }*/
  }
  else
  {
    if (shouldRestore) { newValue = oldValue; }
    else { newValue = psp->simulate(args,trace->getRNG()); } // TODO rng
  }

  if (dynamic_cast<RequestNode*>(node)) { assert(dynamic_pointer_cast<VentureRequest>(newValue)); }

  trace->setValue(node,newValue);

  psp->incorporate(newValue,args);

  if (dynamic_pointer_cast<VentureSPRecord>(newValue)) { processMadeSP(trace,node,scaffold->isAAA(node),shouldRestore,db); }
  if (psp->isRandom()) { trace->registerUnconstrainedChoice(node); }

  if (dynamic_pointer_cast<ScopeIncludeOutputPSP>(psp))
  {
    ScopeID scope = trace->getValue(node->operandNodes[0]);
    BlockID block = trace->getValue(node->operandNodes[1]);
    Node * blockNode = node->operandNodes[2];
    trace->registerUnconstrainedChoiceInScope(scope,block,blockNode);
  }
  return weight;
}


double evalRequests(Trace * trace,
              RequestNode * requestNode,
              shared_ptr<Scaffold> scaffold,
              bool shouldRestore,
              shared_ptr<DB> db,
              shared_ptr<map<Node*,Gradient> > gradients)
{
  //cout << "evalRequests(" << requestNode << "," << requestNode->outputNode << ")" << endl;

  double weight = 0;
  const vector<ESR>& esrs = trace->getValue(requestNode)->getESRs();
  const vector<shared_ptr<LSR> >& lsrs = trace->getValue(requestNode)->getLSRs();

  for (size_t i = 0; i < esrs.size(); ++i)
  {
    const ESR& esr = esrs[i];
    if (!trace->containsMadeSPFamily(trace->getOperatorSPMakerNode(requestNode),esr.id))
    {
      RootOfFamily esrRoot;
      shared_ptr<SP> sp = trace->getMadeSP(trace->getOperatorSPMakerNode(requestNode));
      if (shouldRestore && db->hasESRParent(sp, esr.id))
      {
        esrRoot = db->getESRParent(sp,esr.id);
        weight += restore(trace,esrRoot.get(),scaffold,db,gradients);
      }
      else
      {
        pair<double,Node*> p = evalFamily(trace,esr.exp,esr.env,scaffold,shouldRestore,db,gradients);
        weight += p.first;
        esrRoot = shared_ptr<Node>(p.second);
      }
      if (trace->containsMadeSPFamily(trace->getOperatorSPMakerNode(requestNode),esr.id))
      {
        // evalFamily already registered a family with this id for the
        // operator being applied here, which means a recursive call
        // to the operator issued a request for the same id.
        // Currently, the only way for that it happen is for a
        // recursive memmed function to call itself with the same
        // arguments.
        throw "Recursive mem argument loop detected.";
      }
      trace->registerMadeSPFamily(trace->getOperatorSPMakerNode(requestNode),esr.id,esrRoot);
    }
    RootOfFamily esrRoot = trace->getMadeSPFamilyRoot(trace->getOperatorSPMakerNode(requestNode),esr.id);
    trace->addESREdge(esrRoot,requestNode->outputNode);
    //cout << "numESRParents(" << requestNode->outputNode << ") = " << trace->getESRParents(requestNode->outputNode).size() << endl;
    assert(!trace->getESRParents(requestNode->outputNode).empty());
  }

  /* Next evaluate LSRs. */
  BOOST_FOREACH (shared_ptr<LSR> lsr, lsrs)
  {
    shared_ptr<LatentDB> latentDB;
    Node * makerNode = trace->getOperatorSPMakerNode(requestNode);
    shared_ptr<SP> sp = trace->getMadeSP(makerNode);
    shared_ptr<SPAux> spAux = trace->getMadeSPAux(makerNode);

    if (db->hasLatentDB(makerNode)) { latentDB = db->getLatentDB(makerNode); }

    weight += sp->simulateLatents(spAux,lsr,shouldRestore,latentDB,trace->getRNG());
  }

  return weight;
}



double restore(Trace * trace,
               Node * node,
               shared_ptr<Scaffold> scaffold,
               shared_ptr<DB> db,
               shared_ptr<map<Node*,Gradient> > gradients) 
{
  //cout << "restore(" << node << ")" << endl;

  double weight = 0;
  ConstantNode * constantNode = dynamic_cast<ConstantNode*>(node);
  LookupNode * lookupNode = dynamic_cast<LookupNode*>(node);
  OutputNode * outputNode = dynamic_cast<OutputNode*>(node);

  if (constantNode || (outputNode && outputNode->isFrozen))
  {
    trace->setValue(node, node->exp);
  }
  else if (lookupNode) 
  { 
    weight += regenParents(trace,lookupNode,scaffold,true,db,gradients);
    trace->reconnectLookup(lookupNode);
    trace->setValue(node,trace->getValue(lookupNode->sourceNode));
  }
  else
  {
    assert(outputNode);
    weight += restore(trace,outputNode->operatorNode,scaffold,db,gradients);
    for (size_t i = 0; i < outputNode->operandNodes.size(); ++i)
    {
      weight += restore(trace,outputNode->operandNodes[i],scaffold,db,gradients);
    }
    weight += apply(trace,outputNode->requestNode,outputNode,scaffold,true,db,gradients);
  }
  return weight;
}

